gpytorch/lazy/lazy_evaluated_kernel_tensor.py

```diff
@@ -122,101 +193,97 @@ def diag(self):
-                            "and diag. Expected size {}. Got size {}.".format(
-                                self.__class__.__name__, expected_shape, res.shape
-                            )
-                        )
+        res = super(Kernel, self.kernel).__call__(x1, x2, diag=True, batch_dims=self.batch_dims, **self.params)
 
-                # If we didn't use batch_dims...
+        # Did this Kernel eat the diag option?
+        # If it does not return a LazyEvaluatedKernelTensor, we can call diag on the output
+        if not isinstance(res, LazyEvaluatedKernelTensor):
+            if res.dim() == x1.dim() and res.shape[-2:] == torch.Size((x1.size(-2), x2.size(-2))):
+                res = res.diag()
+
+        # Now we'll make sure that the shape we're getting from diag makes sense
+        if settings.debug.on():
+            # If we used batch_dims...
+            shape = self.kernel.size(x1, x2)
+            if self.batch_dims == (0, 2):
+                if len(shape) == 2:
+                    expected_shape = torch.Size((x1.size(-1), shape[0]))
                 else:
-                    expected_shape = shape[:-1]
-                    if res.shape != expected_shape:
-                        raise RuntimeError(
-                            "The kernel {} is not equipped to handle and diag. Expected size {}. "
-                            "Got size {}".format(self.__class__.__name__, expected_shape, res.shape)
+                    expected_shape = torch.Size((shape[0] * x1.size(-1), shape[1]))
+                if res.shape != expected_shape:
+                    raise RuntimeError(
+                        "The kernel {} is not equipped to handle batch_dims=(0, 2) "
+                        "and diag. Expected size {}. Got size {}.".format(
+                            self.__class__.__name__, expected_shape, res.shape
                         )
+                    )
 
-            if isinstance(res, LazyTensor):
-                res = res.evaluate()
-            self._cached_kernel_diag = res.view(self.shape[:-1]).contiguous()
-            return self._cached_kernel_diag
+            # If we didn't use batch_dims...
+            else:
+                expected_shape = shape[:-1]
+                if res.shape != expected_shape:
+                    raise RuntimeError(
+                        "The kernel {} is not equipped to handle and diag. Expected size {}. "
+                        "Got size {}".format(self.__class__.__name__, expected_shape, res.shape)
+                    )
 
+        if isinstance(res, LazyTensor):
+            res = res.evaluate()
+        return res.view(self.shape[:-1]).contiguous()
+
+    @cached(name="kernel_eval")
     def evaluate_kernel(self):
         """
         NB: This is a meta LazyTensor, in the sense that evaluate can return
         a LazyTensor if the kernel being evaluated does so.
         """
-        if hasattr(self, "_cached_kernel_eval"):
```

Comment:
Was removing this intentional? If so it seems strange to still save the kernel call in a private cache. Also, I think this has perf implications for predictions (see below).

Reply:
using cached annotation